# 1. Import relevant libraries

In [1]:
import numpy as np 

import pandas as pd 
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", None)

import json # library to handle JSON files

from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from bs4 import BeautifulSoup # library to parse HTML and XML documents

from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe


import matplotlib.cm as cm
import matplotlib.colors as colors





import folium

print("Libraries imported.")

Libraries imported.


# 2. Data scraping from Wikipedia

In [2]:
data=requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text

The above is a get request which scrapes text from the webpage

In [3]:
soup=BeautifulSoup(data,'html.parser')

In [4]:
postal_code=[]
borough=[]
neighborhood=[]

### BeautifulSoup functions

In HTML, tables are marked as 'tr'

*  soup.find('table').find_all('tr') finds all tables in the webpage 

In HTML, all table rows are marked as 'td'. Hence, to obtain the data, we need to go through each td instance as shown below

* for rows in soup.find('table').find_all('tr'):


    cells=rows.find_all('td')


In [5]:
for row in soup.find('table').find_all('tr'):
    cells=row.find_all('td')
    if (len(cells)>0):
        postal_code.append(cells[0].text)
        borough.append(cells[1].text)
        neighborhood.append(cells[2].text)

### Lets take a look at the lists we have scrapped in the webpage

In [6]:
postal_code_list=[]
for string in postal_code:
    postal_code_list.append(string.replace('\n',''))
    
postal_code_list[0:10]

['M1A', 'M2A', 'M3A', 'M4A', 'M5A', 'M6A', 'M7A', 'M8A', 'M9A', 'M1B']

In [7]:
borough[0:10]
borough_list=[]

for b in borough:
    borough_list.append(b[0:-1])

borough_list[0:10]

['Not assigned',
 'Not assigned',
 'North York',
 'North York',
 'Downtown Toronto',
 'North York',
 'Downtown Toronto',
 'Not assigned',
 'Etobicoke',
 'Scarborough']

In [8]:
neighborhood[0:10]
neighborhood_list=[]

for n in neighborhood:
    neighborhood_list.append(n[0:-1])
neighborhood_list[0:10]

['Not assigned',
 'Not assigned',
 'Parkwoods',
 'Victoria Village',
 'Regent Park, Harbourfront',
 'Lawrence Manor, Lawrence Heights',
 "Queen's Park, Ontario Provincial Government",
 'Not assigned',
 'Islington Avenue, Humber Valley Village',
 'Malvern, Rouge']

### Now, let us transform the above lists into a pandas dataframe for better readability

# 3. Dataframe creation

In [9]:
df=pd.DataFrame({'Postal code':postal_code_list,'Borough':borough_list,'Neighborhood':neighborhood_list})
df.head()

Postal code           Borough               Neighborhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

 However, we need to pre process the data since we have a lot of "Not assigned" in the Borough and Neighborhood columns

In [10]:
(df['Neighborhood']=='Not assigned').sum()

77

In [11]:
(df['Borough']=='Not assigned').sum()

77

We shall drop all the entries where Borough name is Not assigned. If Borough name is assigned but neighborhood in not, then Borough name is Neighborhood name

In [12]:
df_processed=df.copy()
df_processed=df[df['Borough']!='Not assigned']
df_processed

Postal code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
8           M9A         Etobicoke   
9           M1B       Scarborough   
11          M3B        North York   
12          M4B         East York   
13          M5B  Downtown Toronto   
14          M6B        North York   
17          M9B         Etobicoke   
18          M1C       Scarborough   
20          M3C        North York   
21          M4C         East York   
22          M5C  Downtown Toronto   
23          M6C              York   
26          M9C         Etobicoke   
27          M1E       Scarborough   
30          M4E      East Toronto   
31          M5E  Downtown Toronto   
32          M6E              York   
36          M1G       Scarborough   
39          M4G         East York   
40          M5G  Downtown Toronto   
41          M6G  Downtown Toronto   
45          M1H       Scarborough   
46          M2H        North York   
47          M3H        North York   
48          M4H         East York   
49          M5H  Downtown Toronto   
50          M6H      West Toronto   
54          M1J       Scarborough   
55          M2J        North York   
56          M3J        North York   
57          M4J         East York   
58          M5J  Downtown Toronto   
59          M6J      West Toronto   
63          M1K       Scarborough   
64          M2K        North York   
65          M3K        North York   
66          M4K      East Toronto   
67          M5K  Downtown Toronto   
68          M6K      West Toronto   
72          M1L       Scarborough   
73          M2L        North York   
74          M3L        North York   
75          M4L      East Toronto   
76          M5L  Downtown Toronto   
77          M6L        North York   
80          M9L        North York   
81          M1M       Scarborough   
82          M2M        North York   
83          M3M        North York   
84          M4M      East Toronto   
85          M5M        North York   
86          M6M              York   
89          M9M        North York   
90          M1N       Scarborough   
91          M2N        North York   
92          M3N        North York   
93          M4N   Central Toronto   
94          M5N   Central Toronto   
95          M6N              York   
98          M9N              York   
99          M1P       Scarborough   
100         M2P        North York   
102         M4P   Central Toronto   
103         M5P   Central Toronto   
104         M6P      West Toronto   
107         M9P         Etobicoke   
108         M1R       Scarborough   
109         M2R        North York   
111         M4R   Central Toronto   
112         M5R   Central Toronto   
113         M6R      West Toronto   
114         M7R       Mississauga   
116         M9R         Etobicoke   
117         M1S       Scarborough   
120         M4S   Central Toronto   
121         M5S  Downtown Toronto   
122         M6S      West Toronto   
126         M1T       Scarborough   
129         M4T   Central Toronto   
130         M5T  Downtown Toronto   
135         M1V       Scarborough   
138         M4V   Central Toronto   
139         M5V  Downtown Toronto   
142         M8V         Etobicoke   
143         M9V         Etobicoke   
144         M1W       Scarborough   
147         M4W  Downtown Toronto   
148         M5W  Downtown Toronto   
151         M8W         Etobicoke   
152         M9W         Etobicoke   
153         M1X       Scarborough   
156         M4X  Downtown Toronto   
157         M5X  Downtown Toronto   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                          Neighborhood  
2                                            Parkwoods  
3                                     Vic

In [13]:
(df_processed['Neighborhood']=='Not assigned').any()

False

Hence, it can be seen that after clearing the unassigned boroughs, there are no unassigned neighborhoods

# 4. Grouping the rows with same postal code 

In [14]:
df_grouped=df_processed.groupby(['Postal code','Borough'],as_index=False).agg(lambda x: ', '.join(x))
df_grouped.head()

Postal code      Borough                            Neighborhood
0         M1B  Scarborough                          Malvern, Rouge
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek
2         M1E  Scarborough       Guildwood, Morningside, West Hill
3         M1G  Scarborough                                  Woburn
4         M1H  Scarborough                               Cedarbrae

# 5. Unassigned neighborhoods as Boroughs

Although thankfully we did face the issue of having unassigned neighborhoods after removing all the unassigned Boroughs, we can test the code for taking care of the unassigned neighborhoods as follows.

If in case the neighborhood is unassigned, it must be the same as the borough name. An example has been shown below.

In [15]:
df['Borough'][0]='sample1'
df['Borough'][1]='sample2'

for index,rows in df.iterrows():
    if rows['Neighborhood']=='Not assigned':
        rows['Neighborhood']=rows['Borough']
           

In [16]:
df.head()

Postal code           Borough               Neighborhood
0         M1A           sample1                    sample1
1         M2A           sample2                    sample2
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

# 6. Comparing the dataframe

In [17]:
postal_test_list = ["M5G", "M2H", "M4B", "M1J", "M4G", "M4M", "M1R", "M9V", "M9L", "M5V", "M1B", "M5A"]

df_test=pd.DataFrame(columns=df_grouped.columns)

for pcodes in postal_test_list:
    df_test=df_test.append(df_grouped[df_grouped['Postal code']==pcodes])

In [18]:
df_test.reset_index(inplace=True)

In [19]:
df_test.drop('index',axis=1,inplace=True)

In [20]:
df_test

Postal code           Borough  \
0          M5G  Downtown Toronto   
1          M2H        North York   
2          M4B         East York   
3          M1J       Scarborough   
4          M4G         East York   
5          M4M      East Toronto   
6          M1R       Scarborough   
7          M9V         Etobicoke   
8          M9L        North York   
9          M5V  Downtown Toronto   
10         M1B       Scarborough   
11         M5A  Downtown Toronto   

                                         Neighborhood  
0                                  Central Bay Street  
1                                   Hillcrest Village  
2                     Parkview Hill, Woodbine Gardens  
3                                 Scarborough Village  
4                                             Leaside  
5                                     Studio District  
6                                   Wexford, Maryvale  
7   South Steeles, Silverstone, Humbergate, Jamest...  
8                                       Humber Summit  
9   CN Tower, King and Spadina, Railway Lands, Har...  
10                                     Malvern, Rouge  
11                          Regent Park, Harbourfront

# 7. Merging postal codes with Latitudes and Longitudes

In [21]:
df_lat_lon=pd.read_csv('Geospatial_Coordinates.csv')
df_lat_lon.rename(columns={'Postal Code':'Postal code'},inplace=True)
df_lat_lon.head()

Postal code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

As we can see, the latitudes and longitudes are postal coded uniquely. Hence, we can merge our data frame by postal codes using the pandas merge function.

In [22]:
df_locations=df_grouped.merge(df_lat_lon, on='Postal code')
df_locations.head()

Postal code      Borough                            Neighborhood   Latitude  \
0         M1B  Scarborough                          Malvern, Rouge  43.806686   
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2         M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3         M1G  Scarborough                                  Woburn  43.770992   
4         M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

Let us check specifically for the postal codes which were asked in the assignment.

In [23]:
df_test1=pd.DataFrame(columns=df_locations.columns)

for pcodes in postal_test_list:
    df_test1=df_test1.append(df_locations[df_locations['Postal code']==pcodes],ignore_index=True)

In [24]:
df_test1

Postal code           Borough  \
0          M5G  Downtown Toronto   
1          M2H        North York   
2          M4B         East York   
3          M1J       Scarborough   
4          M4G         East York   
5          M4M      East Toronto   
6          M1R       Scarborough   
7          M9V         Etobicoke   
8          M9L        North York   
9          M5V  Downtown Toronto   
10         M1B       Scarborough   
11         M5A  Downtown Toronto   

                                         Neighborhood   Latitude  Longitude  
0                                  Central Bay Street  43.657952 -79.387383  
1                                   Hillcrest Village  43.803762 -79.363452  
2                     Parkview Hill, Woodbine Gardens  43.706397 -79.309937  
3                                 Scarborough Village  43.744734 -79.239476  
4                                             Leaside  43.709060 -79.363452  
5                                     Studio District  43.659526 -79.340923  
6                                   Wexford, Maryvale  43.750072 -79.295849  
7   South Steeles, Silverstone, Humbergate, Jamest...  43.739416 -79.588437  
8                                       Humber Summit  43.756303 -79.565963  
9   CN Tower, King and Spadina, Railway Lands, Har...  43.628947 -79.394420  
10                                     Malvern, Rouge  43.806686 -79.194353  
11                          Regent Park, Harbourfront  43.654260 -79.360636

# 8. Creation of Toronto map

From the above dataframe, we can estimate the latitudes and logitudes of Toronto. We could use geolocator as suggested by Coursera aswell.

In [25]:
#Toronto coordinates

latitude=43.65
longitude=-79.4

map_toronto=folium.Map(location=[latitude,longitude],zoom_start=11)
map_toronto

In [26]:
for lat,lon,borough,neighborhood in zip(df_locations['Latitude'],df_locations['Longitude'],df_locations['Borough'],df_locations['Neighborhood']):
                                        
                                        label='{} {}'.format(neighborhood,borough)
                                        label=folium.Popup(label)
                                        
                                        folium.CircleMarker(
                                            [lat,lon], radius=5,popup=label,color='orange',fill=True,fill_color='black',fill_opacity=0.6).add_to(map_toronto)
                                   

In [27]:
map_toronto

# 9. Downsizing the dataframe


In [28]:
boroughs=list(df_locations['Borough'].unique())
toronto_boroughs=[]

for names in boroughs:
    if 'toronto' in names.lower():
        toronto_boroughs.append(names)
        
toronto_boroughs

['East Toronto', 'Central Toronto', 'Downtown Toronto', 'West Toronto']

In [29]:
df_toronto=pd.DataFrame(columns=df_locations.columns)

for names in toronto_boroughs:
    df_toronto=df_toronto.append(df_locations[df_locations['Borough']==names],ignore_index=True)
    
df_toronto.head()

Postal code       Borough  \
0         M4E  East Toronto   
1         M4K  East Toronto   
2         M4L  East Toronto   
3         M4M  East Toronto   
4         M7Y  East Toronto   

                                        Neighborhood   Latitude  Longitude  
0                                        The Beaches  43.676357 -79.293031  
1                       The Danforth West, Riverdale  43.679557 -79.352188  
2                     India Bazaar, The Beaches West  43.668999 -79.315572  
3                                    Studio District  43.659526 -79.340923  
4  Business reply mail Processing Centre, South C...  43.662744 -79.321558

# 10. Using foursquare for location exploration

In [30]:
CLIENT_ID='ZALY0GYPRA1XXX0BZZQNLFG1C3FBIFA4E23TL4XPUSQ42G1P'
CLIENT_SECRET='XA524DRFQCWEIRUGCIHPVUOADC32EPK3OPCHFRSUERFXCPWH'
VERSION = '20180605'

Let us explore the top 10 places within a 500 meter radius

In [31]:
radius = 500
LIMIT = 10

venues = []

for lat, lon, pcode, borough, neighborhood in zip(df_toronto['Latitude'], df_toronto['Longitude'], df_toronto['Postal code'], df_toronto['Borough'], df_toronto['Neighborhood']):
    url = "https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}".format(
        CLIENT_ID,
        CLIENT_SECRET,
        VERSION,
        lat,
        lon,
        radius, 
        LIMIT)
    
    results = requests.get(url).json()["response"]['groups'][0]['items']
    
    for venue in results:
        venues.append((
            pcode, 
            borough,
            neighborhood,
            lat, 
            lon, 
            venue['venue']['name'], 
            venue['venue']['location']['lat'], 
            venue['venue']['location']['lng'],  
            venue['venue']['categories'][0]['name']))


In [32]:
venues_df=pd.DataFrame(venues)
venues_df.head()

0             1                             2          3          4  \
0  M4E  East Toronto                   The Beaches  43.676357 -79.293031   
1  M4E  East Toronto                   The Beaches  43.676357 -79.293031   
2  M4E  East Toronto                   The Beaches  43.676357 -79.293031   
3  M4E  East Toronto                   The Beaches  43.676357 -79.293031   
4  M4K  East Toronto  The Danforth West, Riverdale  43.679557 -79.352188   

                                    5          6          7                  8  
0                   Glen Manor Ravine  43.676821 -79.293942              Trail  
1  The Big Carrot Natural Food Market  43.678879 -79.297734  Health Food Store  
2                 Grover Pub and Grub  43.679181 -79.297215                Pub  
3                       Upper Beaches  43.680563 -79.292869       Neighborhood  
4                       MenEssentials  43.677820 -79.351265     Cosmetics Shop

In [33]:
venues_df.rename(columns={0:'Postal code',1:'Borough',2:'Neighborhood',3:'Borough latitude',4:'Borough longitude',5:'Venue name',6:'Venue latitude',7:'Venue longitude',8:'Category'},inplace=True)
venues_df.head()

Postal code       Borough                  Neighborhood  Borough latitude  \
0         M4E  East Toronto                   The Beaches         43.676357   
1         M4E  East Toronto                   The Beaches         43.676357   
2         M4E  East Toronto                   The Beaches         43.676357   
3         M4E  East Toronto                   The Beaches         43.676357   
4         M4K  East Toronto  The Danforth West, Riverdale         43.679557   

   Borough longitude                          Venue name  Venue latitude  \
0         -79.293031                   Glen Manor Ravine       43.676821   
1         -79.293031  The Big Carrot Natural Food Market       43.678879   
2         -79.293031                 Grover Pub and Grub       43.679181   
3         -79.293031                       Upper Beaches       43.680563   
4         -79.352188                       MenEssentials       43.677820   

   Venue longitude           Category  
0       -79.293942              Trail  
1       -79.297734  Health Food Store  
2       -79.297215                Pub  
3       -79.292869       Neighborhood  
4       -79.351265     Cosmetics Shop

# 11. Searching for a specific venue category

In [37]:
print('Number of unique categories of venues : {}'.format(len(venues_df)))

Number of unique categories of venues : 348


As we can see, there are about 348 venue categories which are a lot. If we are interested in a Steakhouse, we can search for a particular query as shown below:

In [50]:
search_query='Steakhouse'
url = 'https://api.foursquare.com/v2/venues/search?client_id={}&client_secret={}&ll={},{}&v={}&query={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION, search_query, radius, LIMIT)

In [51]:
results=requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5ede6b7cfb34b5001b486051'},
 'response': {'venues': [{'id': '54256356498edef2967366eb',
    'name': 'The Keg Steakhouse + Bar - King West',
    'location': {'address': '560 King St W',
     'lat': 43.64477859109389,
     'lng': -79.3992247945535,
     'labeledLatLngs': [{'label': 'display',
       'lat': 43.64477859109389,
       'lng': -79.3992247945535}],
     'distance': 584,
     'postalCode': 'M5V 0L5',
     'cc': 'CA',
     'city': 'Toronto',
     'state': 'ON',
     'country': 'Canada',
     'formattedAddress': ['560 King St W', 'Toronto ON M5V 0L5', 'Canada']},
    'categories': [{'id': '4bf58dd8d48988d1c4941735',
      'name': 'Restaurant',
      'pluralName': 'Restaurants',
      'shortName': 'Restaurant',
      'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/default_',
       'suffix': '.png'},
      'primary': True}],
    'referralId': 'v-1591634989',
    'hasPerk': False},
   {'id': '55fedf4c498e47bc0c154f91',
    'name'

In [52]:
# assign relevant part of JSON to venues
venues = results['response']['venues']

# tranform venues into a dataframe
dataframe = json_normalize(venues)
dataframe.head()

id                                  name  \
0  54256356498edef2967366eb  The Keg Steakhouse + Bar - King West   
1  55fedf4c498e47bc0c154f91              The Keg Steakhouse + Bar   
2  4b22f410f964a520005124e3                          Jacobs & Co.   

                                          categories    referralId  hasPerk  \
0  [{'id': '4bf58dd8d48988d1c4941735', 'name': 'R...  v-1591634989    False   
1  [{'id': '4bf58dd8d48988d1cc941735', 'name': 'S...  v-1591634989    False   
2  [{'id': '4bf58dd8d48988d1cc941735', 'name': 'S...  v-1591634989    False   

       location.address  location.lat  location.lng  \
0         560 King St W     43.644779    -79.399225   
1  560 King Street West     43.646760    -79.395368   
2           12 Brant St     43.645339    -79.398020   

                             location.labeledLatLngs  location.distance  \
0  [{'label': 'display', 'lat': 43.64477859109389...                584   
1  [{'label': 'display', 'lat': 43.64676043755301...                518   
2  [{'label': 'display', 'lat': 43.64533925830643...                542   

  location.postalCode location.cc location.city location.state  \
0             M5V 0L5          CA       Toronto             ON   
1             M5V 0L5          CA       Toronto             ON   
2             M5V 2M1          CA       Toronto             ON   

  location.country                          location.formattedAddress  \
0           Canada        [560 King St W, Toronto ON M5V 0L5, Canada]   
1           Canada  [560 King Street West, Toronto ON M5V 0L5, Can...   
2           Canada   [12 Brant St (King), Toronto ON M5V 2M1, Canada]   

  location.crossStreet  
0                  NaN  
1                  NaN  
2                 King

# 12. Data analysis

Let us check the various categories of venues returned

In [34]:
venues_df['Category'].unique()

array(['Trail', 'Health Food Store', 'Pub', 'Neighborhood',
       'Cosmetics Shop', 'Greek Restaurant', 'Italian Restaurant',
       'Ice Cream Shop', 'Yoga Studio', 'Brewery',
       'Fruit & Vegetable Store', 'Gym', 'Fish & Chips Shop',
       'Fast Food Restaurant', 'Sushi Restaurant', 'Park', 'Liquor Store',
       'Pet Store', 'Bookstore', 'Fish Market', 'Sandwich Place',
       'Bakery', 'Coffee Shop', 'Gay Bar', 'Farmers Market', 'Comic Shop',
       'Burrito Place', 'Pizza Place', 'Restaurant', 'Skate Park',
       'Garden Center', 'Auto Workshop', 'Swim School', 'Bus Line',
       'Breakfast Spot', 'Food & Drink Shop', 'Department Store', 'Hotel',
       'Gym / Fitness Center', 'Diner', 'Salon / Barbershop',
       'Mexican Restaurant', 'Spa', 'Clothing Store',
       'Chinese Restaurant', 'Dessert Shop', 'Café', 'Indian Restaurant',
       'Seafood Restaurant', 'Summer Camp', 'Tennis Court', 'Lawyer',
       'Supermarket', 'American Restaurant', 'Sports Bar',
       'Fried C

Let us check the number of venues returned for each postal code

In [36]:
venues_df.groupby(['Postal code','Borough','Neighborhood']).count()

Borough latitude  \
Postal code Borough          Neighborhood                                                           
M4E         East Toronto     The Beaches                                                        4   
M4K         East Toronto     The Danforth West, Riverdale                                      10   
M4L         East Toronto     India Bazaar, The Beaches West                                    10   
M4M         East Toronto     Studio District                                                   10   
M4N         Central Toronto  Lawrence Park                                                      3   
M4P         Central Toronto  Davisville North                                                   8   
M4R         Central Toronto  North Toronto West,  Lawrence Park                                10   
M4S         Central Toronto  Davisville                                                        10   
M4T         Central Toronto  Moore Park, Summerhill East                                        4   
M4V         Central Toronto  Summerhill West, Rathnelly, South Hill, Forest ...                10   
M4W         Downtown Toronto Rosedale                                                           4   
M4X         Downtown Toronto St. James Town, Cabbagetown                                       10   
M4Y         Downtown Toronto Church and Wellesley                                              10   
M5A         Downtown Toronto Regent Park, Harbourfront                                         10   
M5B         Downtown Toronto Garden District, Ryerson                                          10   
M5C         Downtown Toronto St. James Town                                                    10   
M5E         Downtown Toronto Berczy Park                                                       10   
M5G         Downtown Toronto Central Bay Street                                                10   
M5H         Downtown Toronto Richmond, Adelaide, King                                          10   
M5J         Downtown Toronto Harbourfront East, Union Station, Toronto Islands                 10   
M5K         Downtown Toronto Toronto Dominion Centre, Design Exchange                          10   
M5L         Downtown Toronto Commerce Court, Victoria Hotel                                    10   
M5N         Central Toronto  Roselawn                                                           1   
M5P         Central Toronto  Forest Hill North & West, Forest Hill Road Park                    4   
M5R         Central Toronto  The Annex, North Midtown, Yorkville                               10   
M5S         Downtown Toronto University of Toronto, Harbord                                    10   
M5T         Downtown Toronto Kensington Market, Chinatown, Grange Park                         10   
M5V         Downtown Toronto CN Tower, King and Spadina, Railway Lands, Harb...                10   
M5W         Downtown Toronto Stn A PO Boxes                                                    10   
M5X         Downtown Toronto First Canadian Place, Underground city                            10   
M6G         Downtown Toronto Christie                                                          10   
M6H         West Toronto     Dufferin, Dovercourt Village                                      10   
M6J         West Toronto     Little Portugal, Trinity                                          10   
M6K         West Toronto     Brockton, Parkdale Village, Exhibition Place                      10   
M6P         West Toronto     High Park, The Junction South                                     10   
M6R         West Toronto     Parkdale, Roncesvalles                                            10   
M6S         West Toronto     Runnymede, Swansea                                                10   
M7A         Downtown Toronto Queen's Park, Ontario Provincial Government                       10   
M7Y         East Toronto     Business rep

As we can see from the data above, most postal codes show the top 10 places. Some of the postal codes lack enough venues.

Let us one hot encode the above dataframefor the venue categories field.

In [53]:
toronto_onehot=pd.get_dummies(venues_df['Category'])

In [57]:
venues_df=venues_df.iloc[:,:-1]

In [60]:
venues_df=venues_df.merge(toronto_onehot,on=venues_df.index)
venues_df

key_0 Postal code           Borough  \
0        0         M4E      East Toronto   
1        1         M4E      East Toronto   
2        2         M4E      East Toronto   
3        3         M4E      East Toronto   
4        4         M4K      East Toronto   
5        5         M4K      East Toronto   
6        6         M4K      East Toronto   
7        7         M4K      East Toronto   
8        8         M4K      East Toronto   
9        9         M4K      East Toronto   
10      10         M4K      East Toronto   
11      11         M4K      East Toronto   
12      12         M4K      East Toronto   
13      13         M4K      East Toronto   
14      14         M4L      East Toronto   
15      15         M4L      East Toronto   
16      16         M4L      East Toronto   
17      17         M4L      East Toronto   
18      18         M4L      East Toronto   
19      19         M4L      East Toronto   
20      20         M4L      East Toronto   
21      21         M4L      East Toronto   
22      22         M4L      East Toronto   
23      23         M4L      East Toronto   
24      24         M4M      East Toronto   
25      25         M4M      East Toronto   
26      26         M4M      East Toronto   
27      27         M4M      East Toronto   
28      28         M4M      East Toronto   
29      29         M4M      East Toronto   
30      30         M4M      East Toronto   
31      31         M4M      East Toronto   
32      32         M4M      East Toronto   
33      33         M4M      East Toronto   
34      34         M7Y      East Toronto   
35      35         M7Y      East Toronto   
36      36         M7Y      East Toronto   
37      37         M7Y      East Toronto   
38      38         M7Y      East Toronto   
39      39         M7Y      East Toronto   
40      40         M7Y      East Toronto   
41      41         M7Y      East Toronto   
42      42         M7Y      East Toronto   
43      43         M7Y      East Toronto   
44      44         M4N   Central Toronto   
45      45         M4N   Central Toronto   
46      46         M4N   Central Toronto   
47      47         M4P   Central Toronto   
48      48         M4P   Central Toronto   
49      49         M4P   Central Toronto   
50      50         M4P   Central Toronto   
51      51         M4P   Central Toronto   
52      52         M4P   Central Toronto   
53      53         M4P   Central Toronto   
54      54         M4P   Central Toronto   
55      55         M4R   Central Toronto   
56      56         M4R   Central Toronto   
57      57         M4R   Central Toronto   
58      58         M4R   Central Toronto   
59      59         M4R   Central Toronto   
60      60         M4R   Central Toronto   
61      61         M4R   Central Toronto   
62      62         M4R   Central Toronto   
63      63         M4R   Central Toronto   
64      64         M4R   Central Toronto   
65      65         M4S   Central Toronto   
66      66         M4S   Central Toronto   
67      67         M4S   Central Toronto   
68      68         M4S   Central Toronto   
69      69         M4S   Central Toronto   
70      70         M4S   Central Toronto   
71      71         M4S   Central Toronto   
72      72         M4S   Central Toronto   
73      73         M4S   Central Toronto   
74      74         M4S   Central Toronto   
75      75         M4T   Central Toronto   
76      76         M4T   Central Toronto   
77      77         M4T   Central Toronto   
78      78         M4T   Central Toronto   
79      79         M4V   Central Toronto   
80      80         M4V   Central Toronto   
81      81         M4V   Central Toronto   
82      82         M4V   Central Toronto   
83      83         M4V   Central Toronto   
84      84         M4V   Central Toronto   
85      85         M4V   Central Toronto   
86      86         M4V   Central Toronto   
87      87         M4V   Central Toronto   
88      88         M4V   Central Toronto   
89      89         M5N   Central Toronto   
9

In [61]:
venues_df.drop('key_0',axis=1,inplace=True)
venues_df.head()

Postal code       Borough                Neighborhood_x  Borough latitude  \
0         M4E  East Toronto                   The Beaches         43.676357   
1         M4E  East Toronto                   The Beaches         43.676357   
2         M4E  East Toronto                   The Beaches         43.676357   
3         M4E  East Toronto                   The Beaches         43.676357   
4         M4K  East Toronto  The Danforth West, Riverdale         43.679557   

   Borough longitude                          Venue name  Venue latitude  \
0         -79.293031                   Glen Manor Ravine       43.676821   
1         -79.293031  The Big Carrot Natural Food Market       43.678879   
2         -79.293031                 Grover Pub and Grub       43.679181   
3         -79.293031                       Upper Beaches       43.680563   
4         -79.352188                       MenEssentials       43.677820   

   Venue longitude  Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0       -79.293942        0                   0             0               0   
1       -79.297734        0                   0             0               0   
2       -79.297215        0                   0             0               0   
3       -79.292869        0                   0             0               0   
4       -79.351265        0                   0             0               0   

   Airport Service  Airport Terminal  American Restaurant  \
0                0                 0                    0   
1                0                 0                    0   
2                0                 0                    0   
3                0                 0                    0   
4                0                 0                    0   

   Arts & Crafts Store  Asian Restaurant  Auto Workshop  BBQ Joint  Bakery  \
0                    0                 0              0          0       0   
1                    0                 0              0          0       0   
2                    0                 0              0          0       0   
3                    0                 0              0          0       0   
4                    0                 0              0          0       0   

   Bank  Bar  Beer Bar  Beer Store  Bookstore  Breakfast Spot  Brewery  \
0     0    0         0           0          0               0        0   
1     0    0         0           0          0               0        0   
2     0    0         0           0          0               0        0   
3     0    0         0           0          0               0        0   
4     0    0         0           0          0               0        0   

   Bubble Tea Shop  Burger Joint  Burrito Place  Bus Line  Café  Candy Store  \
0                0             0              0         0     0            0   
1                0             0              0         0     0            0   
2                0             0              0         0     0            0   
3                0             0              0         0     0            0   
4                0             0              0         0     0            0   

   Chinese Restaurant  Clothing Store  Cocktail Bar  Coffee Shop  Comic Shop  \
0                   0               0             0            0           0   
1                   0               0             0            0           0   
2                   0               0             0            0           0   
3                   0               0             0            0           0   
4                   0               0             0            0           0   

   Concert Hall  Cosmetics Shop  Creperie  Cuban Restaurant  Dance Studio  \
0             0               0         0                 0             0   
1             0               0         0                 0             0   
2             0               0         0                 0             0   
3             0               0 

In [67]:
venues_df.rename(columns={'Neighborhood_x':'Neighborhood'},inplace=True)
venues_df.head()

Postal code       Borough                  Neighborhood  Borough latitude  \
0         M4E  East Toronto                   The Beaches         43.676357   
1         M4E  East Toronto                   The Beaches         43.676357   
2         M4E  East Toronto                   The Beaches         43.676357   
3         M4E  East Toronto                   The Beaches         43.676357   
4         M4K  East Toronto  The Danforth West, Riverdale         43.679557   

   Borough longitude                          Venue name  Venue latitude  \
0         -79.293031                   Glen Manor Ravine       43.676821   
1         -79.293031  The Big Carrot Natural Food Market       43.678879   
2         -79.293031                 Grover Pub and Grub       43.679181   
3         -79.293031                       Upper Beaches       43.680563   
4         -79.352188                       MenEssentials       43.677820   

   Venue longitude  Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0       -79.293942        0                   0             0               0   
1       -79.297734        0                   0             0               0   
2       -79.297215        0                   0             0               0   
3       -79.292869        0                   0             0               0   
4       -79.351265        0                   0             0               0   

   Airport Service  Airport Terminal  American Restaurant  \
0                0                 0                    0   
1                0                 0                    0   
2                0                 0                    0   
3                0                 0                    0   
4                0                 0                    0   

   Arts & Crafts Store  Asian Restaurant  Auto Workshop  BBQ Joint  Bakery  \
0                    0                 0              0          0       0   
1                    0                 0              0          0       0   
2                    0                 0              0          0       0   
3                    0                 0              0          0       0   
4                    0                 0              0          0       0   

   Bank  Bar  Beer Bar  Beer Store  Bookstore  Breakfast Spot  Brewery  \
0     0    0         0           0          0               0        0   
1     0    0         0           0          0               0        0   
2     0    0         0           0          0               0        0   
3     0    0         0           0          0               0        0   
4     0    0         0           0          0               0        0   

   Bubble Tea Shop  Burger Joint  Burrito Place  Bus Line  Café  Candy Store  \
0                0             0              0         0     0            0   
1                0             0              0         0     0            0   
2                0             0              0         0     0            0   
3                0             0              0         0     0            0   
4                0             0              0         0     0            0   

   Chinese Restaurant  Clothing Store  Cocktail Bar  Coffee Shop  Comic Shop  \
0                   0               0             0            0           0   
1                   0               0             0            0           0   
2                   0               0             0            0           0   
3                   0               0             0            0           0   
4                   0               0             0            0           0   

   Concert Hall  Cosmetics Shop  Creperie  Cuban Restaurant  Dance Studio  \
0             0               0         0                 0             0   
1             0               0         0                 0             0   
2             0               0         0                 0             0   
3             0               0 

We see that one postal code has various entries making the dataframe extremely large. Let us downsize the dataframe using groupby method.

In [122]:
venues_grouped=venues_df.groupby(['Postal code','Borough','Neighborhood']).mean().reset_index()

In [123]:
venues_grouped.head()

Postal code          Borough                    Neighborhood  \
0         M4E     East Toronto                     The Beaches   
1         M4K     East Toronto    The Danforth West, Riverdale   
2         M4L     East Toronto  India Bazaar, The Beaches West   
3         M4M     East Toronto                 Studio District   
4         M4N  Central Toronto                   Lawrence Park   

   Borough latitude  Borough longitude  Venue latitude  Venue longitude  \
0         43.676357         -79.293031       43.678861       -79.295440   
1         43.679557         -79.352188       43.677804       -79.350938   
2         43.668998         -79.315572       43.667437       -79.315405   
3         43.659525         -79.340923       43.661016       -79.340683   
4         43.728020         -79.388790       43.727840       -79.386683   

   Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0      0.0                 0.0           0.0             0.0              0.0   
1      0.0                 0.0           0.0             0.0              0.0   
2      0.0                 0.0           0.0             0.0              0.0   
3      0.0                 0.0           0.0             0.0              0.0   
4      0.0                 0.0           0.0             0.0              0.0   

   Airport Terminal  American Restaurant  Arts & Crafts Store  \
0               0.0                  0.0                  0.0   
1               0.0                  0.0                  0.0   
2               0.0                  0.0                  0.0   
3               0.0                  0.0                  0.0   
4               0.0                  0.0                  0.0   

   Asian Restaurant  Auto Workshop  BBQ Joint  Bakery  Bank  Bar  Beer Bar  \
0               0.0            0.0        0.0     0.0   0.0  0.0       0.0   
1               0.0            0.0        0.0     0.0   0.0  0.0       0.0   
2               0.0            0.0        0.0     0.0   0.0  0.0       0.0   
3               0.0            0.0        0.0     0.1   0.0  0.0       0.0   
4               0.0            0.0        0.0     0.0   0.0  0.0       0.0   

   Beer Store  Bookstore  Breakfast Spot  Brewery  Bubble Tea Shop  \
0         0.0        0.0             0.0      0.0              0.0   
1         0.0        0.0             0.0      0.1              0.0   
2         0.0        0.0             0.0      0.1              0.0   
3         0.0        0.1             0.0      0.0              0.0   
4         0.0        0.0             0.0      0.0              0.0   

   Burger Joint  Burrito Place  Bus Line  Café  Candy Store  \
0           0.0            0.0  0.000000   0.0          0.0   
1           0.0            0.0  0.000000   0.0          0.0   
2           0.0            0.0  0.000000   0.0          0.0   
3           0.0            0.0  0.000000   0.0          0.0   
4           0.0            0.0  0.333333   0.0          0.0   

   Chinese Restaurant  Clothing Store  Cocktail Bar  Coffee Shop  Comic Shop  \
0                 0.0             0.0           0.0          0.0         0.0   
1                 0.0             0.0           0.0          0.0         0.0   
2                 0.0             0.0           0.0          0.0         0.0   
3                 0.0             0.0           0.0          0.1         0.0   
4                 0.0             0.0           0.0          0.0         0.0   

   Concert Hall  Cosmetics Shop  Creperie  Cuban Restaurant  Dance Studio  \
0           0.0             0.0       0.0               0.0           0.0   
1           0.0             0.1       0.0               0.0           0.0   
2           0.0             0.0       0.0               0.0           0.0   
3           0.0             0.0       0.0               0.0           0.0   
4           0.0             0.0       0.0               0.0           0.0   

   Department Store  Dessert Shop  Diner  Distribution Center  

Let us check the concised dataframe size.

In [124]:
venues_grouped.shape

(39, 130)

Let us remove the latitudes and longitudes to simplify the dataframe further.

In [85]:
venues_grouped.drop(['Borough latitude','Borough longitude','Venue latitude','Venue longitude'],axis=1,inplace=True)

In [86]:
venues_grouped.head()

Postal code          Borough                    Neighborhood  Airport  \
0         M4E     East Toronto                     The Beaches      0.0   
1         M4K     East Toronto    The Danforth West, Riverdale      0.0   
2         M4L     East Toronto  India Bazaar, The Beaches West      0.0   
3         M4M     East Toronto                 Studio District      0.0   
4         M4N  Central Toronto                   Lawrence Park      0.0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                 0.0           0.0             0.0              0.0   
1                 0.0           0.0             0.0              0.0   
2                 0.0           0.0             0.0              0.0   
3                 0.0           0.0             0.0              0.0   
4                 0.0           0.0             0.0              0.0   

   Airport Terminal  American Restaurant  Arts & Crafts Store  \
0               0.0                  0.0                  0.0   
1               0.0                  0.0                  0.0   
2               0.0                  0.0                  0.0   
3               0.0                  0.0                  0.0   
4               0.0                  0.0                  0.0   

   Asian Restaurant  Auto Workshop  BBQ Joint  Bakery  Bank  Bar  Beer Bar  \
0               0.0            0.0        0.0     0.0   0.0  0.0       0.0   
1               0.0            0.0        0.0     0.0   0.0  0.0       0.0   
2               0.0            0.0        0.0     0.0   0.0  0.0       0.0   
3               0.0            0.0        0.0     0.1   0.0  0.0       0.0   
4               0.0            0.0        0.0     0.0   0.0  0.0       0.0   

   Beer Store  Bookstore  Breakfast Spot  Brewery  Bubble Tea Shop  \
0         0.0        0.0             0.0      0.0              0.0   
1         0.0        0.0             0.0      0.1              0.0   
2         0.0        0.0             0.0      0.1              0.0   
3         0.0        0.1             0.0      0.0              0.0   
4         0.0        0.0             0.0      0.0              0.0   

   Burger Joint  Burrito Place  Bus Line  Café  Candy Store  \
0           0.0            0.0  0.000000   0.0          0.0   
1           0.0            0.0  0.000000   0.0          0.0   
2           0.0            0.0  0.000000   0.0          0.0   
3           0.0            0.0  0.000000   0.0          0.0   
4           0.0            0.0  0.333333   0.0          0.0   

   Chinese Restaurant  Clothing Store  Cocktail Bar  Coffee Shop  Comic Shop  \
0                 0.0             0.0           0.0          0.0         0.0   
1                 0.0             0.0           0.0          0.0         0.0   
2                 0.0             0.0           0.0          0.0         0.0   
3                 0.0             0.0           0.0          0.1         0.0   
4                 0.0             0.0           0.0          0.0         0.0   

   Concert Hall  Cosmetics Shop  Creperie  Cuban Restaurant  Dance Studio  \
0           0.0             0.0       0.0               0.0           0.0   
1           0.0             0.1       0.0               0.0           0.0   
2           0.0             0.0       0.0               0.0           0.0   
3           0.0             0.0       0.0               0.0           0.0   
4           0.0             0.0       0.0               0.0           0.0   

   Department Store  Dessert Shop  Diner  Distribution Center  Dog Run  \
0               0.0           0.0    0.0                  0.0      0.0   
1               0.0           0.0    0.0                  0.0      0.0   
2               0.0           0.0    0.0                  0.0      0.0   
3               0.0           0.0    0.0                  0.0      0.0   
4               0.0           0.0    0.0                  0.0      0.0   

   Donut Shop  Eastern European Restaurant  Falafel Restaurant  \
0  

Let us find out the top 5 venue categories of every postal code, borough and neighborhood as shown below.

In [106]:
num_top_venues=5

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
areaColumns = ['Postal code', 'Borough', 'Neighborhood']
freqColumns = []
for ind in np.arange(num_top_venues):
    try:
        freqColumns.append('{}{} Most common venue'.format(ind+1, indicators[ind]))
    except:
        freqColumns.append('{}th Most common venue'.format(ind+1))
columns = areaColumns+freqColumns

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Postal code'] = venues_grouped['Postal code']
neighborhoods_venues_sorted['Borough'] = venues_grouped['Borough']
neighborhoods_venues_sorted['Neighborhood'] = venues_grouped['Neighborhood']



for ind in np.arange(venues_grouped.shape[0]):
    row_categories=venues_grouped.iloc[ind,:].iloc[3:]
    row_categories_sorted=row_categories.sort_values(ascending=False)
    neighborhoods_venues_sorted.iloc[ind,3:]=row_categories_sorted.index.values[0:num_top_venues]

neighborhoods_venues_sorted.sort_values(freqColumns,inplace=True) 
    
neighborhoods_venues_sorted.reset_index(inplace=True,drop=True)
neighborhoods_venues_sorted

Postal code           Borough  \
0          M5V  Downtown Toronto   
1          M6H      West Toronto   
2          M4M      East Toronto   
3          M5A  Downtown Toronto   
4          M5L  Downtown Toronto   
5          M5K  Downtown Toronto   
6          M4X  Downtown Toronto   
7          M6G  Downtown Toronto   
8          M5R   Central Toronto   
9          M5B  Downtown Toronto   
10         M6S      West Toronto   
11         M5T  Downtown Toronto   
12         M5W  Downtown Toronto   
13         M5X  Downtown Toronto   
14         M5C  Downtown Toronto   
15         M5G  Downtown Toronto   
16         M6K      West Toronto   
17         M4V   Central Toronto   
18         M5E  Downtown Toronto   
19         M7Y      East Toronto   
20         M4L      East Toronto   
21         M5N   Central Toronto   
22         M6P      West Toronto   
23         M6R      West Toronto   
24         M4K      East Toronto   
25         M4E      East Toronto   
26         M4S   Central Toronto   
27         M5P   Central Toronto   
28         M4T   Central Toronto   
29         M5H  Downtown Toronto   
30         M4N   Central Toronto   
31         M5J  Downtown Toronto   
32         M4Y  Downtown Toronto   
33         M4W  Downtown Toronto   
34         M6J      West Toronto   
35         M4P   Central Toronto   
36         M7A  Downtown Toronto   
37         M5S  Downtown Toronto   
38         M4R   Central Toronto   

                                         Neighborhood 1st Most common venue  \
0   CN Tower, King and Spadina, Railway Lands, Har...        Airport Lounge   
1                        Dufferin, Dovercourt Village                Bakery   
2                                     Studio District             Bookstore   
3                           Regent Park, Harbourfront        Breakfast Spot   
4                      Commerce Court, Victoria Hotel                  Café   
5            Toronto Dominion Centre, Design Exchange                  Café   
6                         St. James Town, Cabbagetown                  Café   
7                                            Christie                  Café   
8                 The Annex, North Midtown, Yorkville                  Café   
9                            Garden District, Ryerson                  Café   
10                                 Runnymede, Swansea                  Café   
11          Kensington Market, Chinatown, Grange Park                  Café   
12                                     Stn A PO Boxes          Cocktail Bar   
13             First Canadian Place, Underground city           Coffee Shop   
14                                     St. James Town           Coffee Shop   
15                                 Central Bay Street           Coffee Shop   
16       Brockton, Parkdale Village, Exhibition Place           Coffee Shop   
17  Summerhill West, Rathnelly, South Hill, Forest...           Coffee Shop   
18                                        Berczy Park          Concert Hall   
19  Business reply mail Processing Centre, South C...        Farmers Market   
20                     India Bazaar, The Beaches West  Fast Food Restaurant   
21                                           Roselawn                Garden   
22                      High Park, The Junction South             Gastropub   
23                             Parkdale, Roncesvalles             Gift Shop   
24                       The Danforth West, Riverdale      Greek Restaurant   
25                                        The Beaches     Health Food Store   
26                                         Davisville    Italian Restaurant   
27    Forest Hill North & West, Forest Hill Road Park         Jewelry Store   
28                        Moore Park, Summerhill East                Lawyer   
29                           Richmond, Adelaide, King        Neighborhood_y   
30                                      Lawrence Park                  Park   
31  Harbourfront East, Union Stati

# 13. Clustering the venues into labels

Let us use 5 clusters to label the data.

In [111]:
from sklearn.cluster import KMeans
K=5

cluster_df=venues_grouped.drop(['Postal code', 'Borough', 'Neighborhood'],axis=1)
cluster_df.head()

Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0      0.0                 0.0           0.0             0.0              0.0   
1      0.0                 0.0           0.0             0.0              0.0   
2      0.0                 0.0           0.0             0.0              0.0   
3      0.0                 0.0           0.0             0.0              0.0   
4      0.0                 0.0           0.0             0.0              0.0   

   Airport Terminal  American Restaurant  Arts & Crafts Store  \
0               0.0                  0.0                  0.0   
1               0.0                  0.0                  0.0   
2               0.0                  0.0                  0.0   
3               0.0                  0.0                  0.0   
4               0.0                  0.0                  0.0   

   Asian Restaurant  Auto Workshop  BBQ Joint  Bakery  Bank  Bar  Beer Bar  \
0               0.0            0.0        0.0     0.0   0.0  0.0       0.0   
1               0.0            0.0        0.0     0.0   0.0  0.0       0.0   
2               0.0            0.0        0.0     0.0   0.0  0.0       0.0   
3               0.0            0.0        0.0     0.1   0.0  0.0       0.0   
4               0.0            0.0        0.0     0.0   0.0  0.0       0.0   

   Beer Store  Bookstore  Breakfast Spot  Brewery  Bubble Tea Shop  \
0         0.0        0.0             0.0      0.0              0.0   
1         0.0        0.0             0.0      0.1              0.0   
2         0.0        0.0             0.0      0.1              0.0   
3         0.0        0.1             0.0      0.0              0.0   
4         0.0        0.0             0.0      0.0              0.0   

   Burger Joint  Burrito Place  Bus Line  Café  Candy Store  \
0           0.0            0.0  0.000000   0.0          0.0   
1           0.0            0.0  0.000000   0.0          0.0   
2           0.0            0.0  0.000000   0.0          0.0   
3           0.0            0.0  0.000000   0.0          0.0   
4           0.0            0.0  0.333333   0.0          0.0   

   Chinese Restaurant  Clothing Store  Cocktail Bar  Coffee Shop  Comic Shop  \
0                 0.0             0.0           0.0          0.0         0.0   
1                 0.0             0.0           0.0          0.0         0.0   
2                 0.0             0.0           0.0          0.0         0.0   
3                 0.0             0.0           0.0          0.1         0.0   
4                 0.0             0.0           0.0          0.0         0.0   

   Concert Hall  Cosmetics Shop  Creperie  Cuban Restaurant  Dance Studio  \
0           0.0             0.0       0.0               0.0           0.0   
1           0.0             0.1       0.0               0.0           0.0   
2           0.0             0.0       0.0               0.0           0.0   
3           0.0             0.0       0.0               0.0           0.0   
4           0.0             0.0       0.0               0.0           0.0   

   Department Store  Dessert Shop  Diner  Distribution Center  Dog Run  \
0               0.0           0.0    0.0                  0.0      0.0   
1               0.0           0.0    0.0                  0.0      0.0   
2               0.0           0.0    0.0                  0.0      0.0   
3               0.0           0.0    0.0                  0.0      0.0   
4               0.0           0.0    0.0                  0.0      0.0   

   Donut Shop  Eastern European Restaurant  Falafel Restaurant  \
0         0.0                          0.0                 0.0   
1         0.0                          0.0                 0.0   
2         0.0                          0.0                 0.0   
3         0.0                          0.0                 0.0   
4         0.0                          0.0                 0.0   

   Farmers Market  Fast Food Restaurant  Fish & Chips Shop  Fish Marke

In order to apply any machine learning, we need to take care of all categorical data. The data frame must have only numerical data and must be properly one hot encoded.

In [114]:
kmeans=KMeans(n_clusters=K).fit(cluster_df)
kmeans

KMeans(algorithm='auto', copy_x=True, init='k-means++', max_iter=300,
       n_clusters=5, n_init=10, n_jobs=None, precompute_distances='auto',
       random_state=None, tol=0.0001, verbose=0)

In [119]:
kmeans.labels_

array([0, 4, 4, 4, 0, 3, 4, 1, 1, 1, 0, 1, 3, 3, 1, 4, 3, 1, 3, 3, 1, 1,
       2, 0, 1, 4, 1, 1, 3, 1, 1, 1, 4, 1, 4, 1, 1, 4, 4], dtype=int32)

Let us fit the labels into a dataframe for better readability

In [121]:
labelled_df=venues_grouped.iloc[:,:3]
labelled_df['Cluster label']=kmeans.labels_
labelled_df

Postal code           Borough  \
0          M4E      East Toronto   
1          M4K      East Toronto   
2          M4L      East Toronto   
3          M4M      East Toronto   
4          M4N   Central Toronto   
5          M4P   Central Toronto   
6          M4R   Central Toronto   
7          M4S   Central Toronto   
8          M4T   Central Toronto   
9          M4V   Central Toronto   
10         M4W  Downtown Toronto   
11         M4X  Downtown Toronto   
12         M4Y  Downtown Toronto   
13         M5A  Downtown Toronto   
14         M5B  Downtown Toronto   
15         M5C  Downtown Toronto   
16         M5E  Downtown Toronto   
17         M5G  Downtown Toronto   
18         M5H  Downtown Toronto   
19         M5J  Downtown Toronto   
20         M5K  Downtown Toronto   
21         M5L  Downtown Toronto   
22         M5N   Central Toronto   
23         M5P   Central Toronto   
24         M5R   Central Toronto   
25         M5S  Downtown Toronto   
26         M5T  Downtown Toronto   
27         M5V  Downtown Toronto   
28         M5W  Downtown Toronto   
29         M5X  Downtown Toronto   
30         M6G  Downtown Toronto   
31         M6H      West Toronto   
32         M6J      West Toronto   
33         M6K      West Toronto   
34         M6P      West Toronto   
35         M6R      West Toronto   
36         M6S      West Toronto   
37         M7A  Downtown Toronto   
38         M7Y      East Toronto   

                                         Neighborhood  Cluster label  
0                                         The Beaches              0  
1                        The Danforth West, Riverdale              4  
2                      India Bazaar, The Beaches West              4  
3                                     Studio District              4  
4                                       Lawrence Park              0  
5                                    Davisville North              3  
6                  North Toronto West,  Lawrence Park              4  
7                                          Davisville              1  
8                         Moore Park, Summerhill East              1  
9   Summerhill West, Rathnelly, South Hill, Forest...              1  
10                                           Rosedale              0  
11                        St. James Town, Cabbagetown              1  
12                               Church and Wellesley              3  
13                          Regent Park, Harbourfront              3  
14                           Garden District, Ryerson              1  
15                                     St. James Town              4  
16                                        Berczy Park              3  
17                                 Central Bay Street              1  
18                           Richmond, Adelaide, King              3  
19  Harbourfront East, Union Station, Toronto Islands              3  
20           Toronto Dominion Centre, Design Exchange              1  
21                     Commerce Court, Victoria Hotel              1  
22                                           Roselawn              2  
23    Forest Hill North & West, Forest Hill Road Park              0  
24                The Annex, North Midtown, Yorkville              1  
25                     University of Toronto, Harbord              4  
26          Kensington Market, Chinatown, Grange Park              1  
27  CN Tower, King and Spadina, Railway Lands, Har...              1  
28                                     Stn A PO Boxes              3  
29             First Canadian Place, Underground city              1  
30                                           Christie              1  
31                       Dufferin, Dovercourt Village              1  
32                           Little Portugal, Trinity              4  
33       Brockton, Parkdale Village, Exhibition Place              1  
34                      High Park, The Junction South              4  
35    

We have clustered the various venues into 5 labels. Let us visualise these venues on a folium map of Toronto. We also add the venue latitudes and longitudes to the above dataframe for map plotting

In [125]:
labelled_df['Latitude']=venues_grouped['Venue latitude']
labelled_df['Longitude']=venues_grouped['Venue longitude']
labelled_df.head()

Postal code          Borough                    Neighborhood  Cluster label  \
0         M4E     East Toronto                     The Beaches              0   
1         M4K     East Toronto    The Danforth West, Riverdale              4   
2         M4L     East Toronto  India Bazaar, The Beaches West              4   
3         M4M     East Toronto                 Studio District              4   
4         M4N  Central Toronto                   Lawrence Park              0   

    Latitude  Longitude  
0  43.678861 -79.295440  
1  43.677804 -79.350938  
2  43.667437 -79.315405  
3  43.661016 -79.340683  
4  43.727840 -79.386683

In [137]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=12)

# set color scheme for the clusters
x = np.arange(K)
ys = [i+x+(i*x)**2 for i in range(K)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, pcode, bor, nei, cluster in zip(labelled_df['Latitude'], labelled_df['Longitude'], labelled_df['Postal code'], labelled_df['Borough'], labelled_df['Neighborhood'], labelled_df['Cluster label']):
    label = folium.Popup('{} ({}): {} - Cluster {}'.format(bor, pcode, nei, cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

We see the various clusters marked according to their locations.

# 14. Checking the various dataframes cluster by cluster

## Cluster 1

In [130]:
first_cluster_df=labelled_df.loc[labelled_df['Cluster label'].isin([0])]
first_cluster_df

Postal code           Borough  \
0          M4E      East Toronto   
4          M4N   Central Toronto   
10         M4W  Downtown Toronto   
23         M5P   Central Toronto   

                                       Neighborhood  Cluster label   Latitude  \
0                                       The Beaches              0  43.678861   
4                                     Lawrence Park              0  43.727840   
10                                         Rosedale              0  43.679754   
23  Forest Hill North & West, Forest Hill Road Park              0  43.699872   

    Longitude  
0  -79.295440  
4  -79.386683  
10 -79.377335  
23 -79.408077

## Cluster 2

In [131]:
second_cluster_df=labelled_df.loc[labelled_df['Cluster label'].isin([1])]
second_cluster_df

Postal code           Borough  \
7          M4S   Central Toronto   
8          M4T   Central Toronto   
9          M4V   Central Toronto   
11         M4X  Downtown Toronto   
14         M5B  Downtown Toronto   
17         M5G  Downtown Toronto   
20         M5K  Downtown Toronto   
21         M5L  Downtown Toronto   
24         M5R   Central Toronto   
26         M5T  Downtown Toronto   
27         M5V  Downtown Toronto   
29         M5X  Downtown Toronto   
30         M6G  Downtown Toronto   
31         M6H      West Toronto   
33         M6K      West Toronto   
35         M6R      West Toronto   
36         M6S      West Toronto   

                                         Neighborhood  Cluster label  \
7                                          Davisville              1   
8                         Moore Park, Summerhill East              1   
9   Summerhill West, Rathnelly, South Hill, Forest...              1   
11                        St. James Town, Cabbagetown              1   
14                           Garden District, Ryerson              1   
17                                 Central Bay Street              1   
20           Toronto Dominion Centre, Design Exchange              1   
21                     Commerce Court, Victoria Hotel              1   
24                The Annex, North Midtown, Yorkville              1   
26          Kensington Market, Chinatown, Grange Park              1   
27  CN Tower, King and Spadina, Railway Lands, Har...              1   
29             First Canadian Place, Underground city              1   
30                                           Christie              1   
31                       Dufferin, Dovercourt Village              1   
33       Brockton, Parkdale Village, Exhibition Place              1   
35                             Parkdale, Roncesvalles              1   
36                                 Runnymede, Swansea              1   

     Latitude  Longitude  
7   43.704647 -79.388631  
8   43.689155 -79.384125  
9   43.687648 -79.396557  
11  43.666833 -79.368773  
14  43.656542 -79.380004  
17  43.658421 -79.385060  
20  43.647780 -79.380680  
21  43.647980 -79.380275  
24  43.674543 -79.407563  
26  43.654236 -79.400258  
27  43.631547 -79.395707  
29  43.648669 -79.381292  
30  43.669982 -79.423062  
31  43.669054 -79.440691  
33  43.638393 -79.426454  
35  43.649921 -79.451439  
36  43.650314 -79.483284

## Cluster 3

In [133]:
third_cluster_df=labelled_df.loc[labelled_df['Cluster label'].isin([2])]
third_cluster_df

Postal code          Borough Neighborhood  Cluster label   Latitude  \
22         M5N  Central Toronto     Roselawn              2  43.712189   

    Longitude  
22 -79.411978

## Cluster 4

In [135]:
fourth_cluster_df=labelled_df.loc[labelled_df['Cluster label']==3]
fourth_cluster_df

Postal code           Borough  \
5          M4P   Central Toronto   
12         M4Y  Downtown Toronto   
13         M5A  Downtown Toronto   
16         M5E  Downtown Toronto   
18         M5H  Downtown Toronto   
19         M5J  Downtown Toronto   
28         M5W  Downtown Toronto   

                                         Neighborhood  Cluster label  \
5                                    Davisville North              3   
12                               Church and Wellesley              3   
13                          Regent Park, Harbourfront              3   
16                                        Berczy Park              3   
18                           Richmond, Adelaide, King              3   
19  Harbourfront East, Union Station, Toronto Islands              3   
28                                     Stn A PO Boxes              3   

     Latitude  Longitude  
5   43.711905 -79.391119  
12  43.666288 -79.383265  
13  43.654015 -79.359960  
16  43.646835 -79.373960  
18  43.650173 -79.384718  
19  43.641124 -79.381885  
28  43.647805 -79.375250

## Cluster 5

In [136]:
fifth_cluster_df=labelled_df.loc[labelled_df['Cluster label']==4]
fifth_cluster_df

Postal code           Borough  \
1          M4K      East Toronto   
2          M4L      East Toronto   
3          M4M      East Toronto   
6          M4R   Central Toronto   
15         M5C  Downtown Toronto   
25         M5S  Downtown Toronto   
32         M6J      West Toronto   
34         M6P      West Toronto   
37         M7A  Downtown Toronto   
38         M7Y      East Toronto   

                                         Neighborhood  Cluster label  \
1                        The Danforth West, Riverdale              4   
2                      India Bazaar, The Beaches West              4   
3                                     Studio District              4   
6                  North Toronto West,  Lawrence Park              4   
15                                     St. James Town              4   
25                     University of Toronto, Harbord              4   
32                           Little Portugal, Trinity              4   
34                      High Park, The Junction South              4   
37        Queen's Park, Ontario Provincial Government              4   
38  Business reply mail Processing Centre, South C...              4   

     Latitude  Longitude  
1   43.677804 -79.350938  
2   43.667437 -79.315405  
3   43.661016 -79.340683  
6   43.715059 -79.400213  
15  43.651302 -79.374467  
25  43.662756 -79.403533  
32  43.647175 -79.420034  
34  43.664684 -79.465895  
37  43.662893 -79.385930  
38  43.664277 -79.322027

## Cluster sizes

In [139]:
cluster_size_df=pd.DataFrame(columns=['Cluster label','size'])
clusters=np.arange(1,6)
cluster_size_df['Cluster label']=clusters


In [143]:
sizes=np.array([first_cluster_df.shape[0],second_cluster_df.shape[0],third_cluster_df.shape[0],fourth_cluster_df.shape[0],fifth_cluster_df.shape[0]])

In [145]:
cluster_size_df['size']=sizes

In [146]:
cluster_size_df

Cluster label  size
0              1     4
1              2    17
2              3     1
3              4     7
4              5    10

As we can see, cluster 2 is the biggest while cluster 3 has only 1 entry.